elixir/apps/web/test/web/live/settings/dns_test.exs

```diff
@@ -167,173 +168,355 @@ defmodule Web.Live.Settings.DNSTest do
+  test "returns error when custom type has no addresses", %{
     account: account,
     identity: identity,
     conn: conn
   } do
+    {:ok, lv, _html} =
+      conn
+      |> authorize_conn(identity)
+      |> live(~p"/#{account}/settings/dns")
+
     attrs = %{
       account: %{
         config: %{
           clients_upstream_dns: %{
-            "0" => %{address: ""}
+            type: "custom",
+            addresses: []
           }
         }
       }
     }
 
+    html =
+      lv
+      |> render_change(:change, attrs)
+      |> render_submit()
+
+    assert html =~ "must have at least one custom resolver"
+  end
+
+  test "validates duplicate addresses", %{
+    account: account,
+    identity: identity,
+    conn: conn
+  } do
     {:ok, lv, _html} =
       conn
       |> authorize_conn(identity)
       |> live(~p"/#{account}/settings/dns")
 
-    lv
-    |> form("form", attrs)
-    |> render_submit()
+    assert lv
+           |> form("form", %{
+             account: %{
+               config: %{
+                 clients_upstream_dns: %{
+                   type: "custom",
+                   addresses: %{
+                     "0" => %{"address" => "8.8.8.8"},
+                     "1" => %{"address" => "8.8.8.8"}
+                   }
+                 }
+               }
+             }
+           })
+           |> render_change() =~ "all addresses must be unique"
+  end
+
+  test "validates IP addresses", %{
+    account: account,
+    identity: identity,
+    conn: conn
+  } do
+    {:ok, lv, _html} =
+      conn
+      |> authorize_conn(identity)
+      |> live(~p"/#{account}/settings/dns")
 
     assert lv
-           |> form("form")
-           |> find_inputs() == [
-             "account[config][_persistent_id]",
-             "account[config][clients_upstream_dns][0][_persistent_id]",
-             "account[config][clients_upstream_dns][0][address]",
-             "account[config][clients_upstream_dns][0][protocol]",
-             "account[config][clients_upstream_dns][1][_persistent_id]",
-             "account[config][clients_upstream_dns][1][address]",
-             "account[config][clients_upstream_dns][1][protocol]",
-             "account[config][clients_upstream_dns][2][_persistent_id]",
-             "account[config][clients_upstream_dns][2][address]",
-             "account[config][clients_upstream_dns][2][protocol]",
-             "account[config][clients_upstream_dns_drop][]",
-             "account[config][clients_upstream_dns_sort][]",
-             "account[config][search_domain]"
-           ]
+           |> form("form", %{
+             account: %{
+               config: %{
+                 clients_upstream_dns: %{
+                   type: "custom",
+                   addresses: %{"0" => %{"address" => "invalid"}}
+                 }
+               }
+             }
+           })
+           |> render_change() =~ "must be a valid IP address"
+
+    refute lv
+           |> form("form", %{
```

Comment:
I know it was like this before, but using the form's id here for the selector might be a little more robust in case we have more than one form in the DOM.